In [1]:
# data load
import pandas as pd
data = pd.read_csv("C:/nh/자연어처리.csv", low_memory=False)
data.head(2)

id        제품명       리뷰  품종  별점    무게
0  a1  만세보령삼광골드미  맛있고 조아요  삼광   5  10kg
1  a1  만세보령삼광골드미  잘 받았습니다  삼광   5  10kg

In [21]:
# 결측치 확인
data = data.fillna('')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 587 entries, 0 to 586
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      587 non-null    object
 1   제품명     587 non-null    object
 2   리뷰      587 non-null    object
 3   품종      587 non-null    object
 4   별점      587 non-null    int64 
 5   무게      587 non-null    object
dtypes: int64(1), object(5)
memory usage: 27.6+ KB


In [5]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf_matrix = TfidfVectorizer().fit_transform(data['리뷰'])
print(tfidf_matrix.shape)

(587, 3464)


리뷰 칼럼에 대해서 tf-idf를 수행한 결과, 587개의 리뷰를 표현하기 위해 총 3464개의 단어가 사용되었음을 보여주고 있다.
이제 코사인 유사도를 사용하면 바로 문서의 유사도를 구할 수 있음.

In [19]:
# TF-IDF 클래스로 BOW 만들기
tfidfv = TfidfVectorizer().fit(data['리뷰'])
print(tfidfv.transform(data['리뷰']).toarray())
print(tfidfv.vocabulary_)
print(tfidfv.vocabulary_)

[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]
{'맛있고': 1025, '조아요': 2769, '받았습니다': 1318, '만세': 956, '보령': 1500, '삼광미': 1689, '보령쌀': 1503, '재구매': 2662, '합니다': 3341, '밥이': 1371, '맛나요': 1005, '맛있어서': 1039, '재구입': 2675, '중간': 2879, '사이즈도': 1667, '있었다면': 2595, '좋았을': 2808, '같아요': 177, '2가지': 53, '뿐이라': 1627, '안성쌀이맛나요ㅡ재구매해요': 2156, '제일맛있는': 2756, '쌀이예요': 2015, '깨진쌀알도없고': 491, '찰지고맛있어요': 2974, '포장': 3193, '배송': 1401, '모두': 1190, '좋네요': 2785, '쌀도': 1944, '믿음이': 1271, '가고': 129, '좋아요': 2796, '앞으로': 2224, '구매하고': 328, '싶습니다': 1909, '빠르게': 1606, '받았어요': 1319, '만세보령쌀': 959, 'gold': 92, '정말': 2743, '맛있습니다': 1036, '먹던쌀이예요': 1108, '아직안뜯어봤는데': 2129, '좋을걸로예상하고있습니다': 2826, '빠른배송': 1609, '감사합니다': 154, '그냥': 401, '좋은데': 2817, '배송잊': 1430, '다른곳으로': 665, '배송문제만': 1412, '빼면': 1620, '괜찮네요': 299, '적극추천': 2710, '배송이': 1429, '빠릅니다': 1611, '빠른': 1608, '보관통이': 1480, '맘에': 1001, '들어요': 861, '등급이': 875, '보통이라

In [15]:
# 비교를 위해 CountVectorizer로 동일한 코퍼스를 벡터화
from sklearn.feature_extraction.text import CountVectorizer
corpus = data['리뷰']
vector = CountVectorizer()
print(vector.fit_transform(corpus).toarray()) # 코퍼스로부터 각 단어의 빈도 수를 기록한다.
print(vector.vocabulary_) # 각 단어의 인덱스가 어떻게 부여되었는지를 보여준다.

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
{'맛있고': 1025, '조아요': 2769, '받았습니다': 1318, '만세': 956, '보령': 1500, '삼광미': 1689, '보령쌀': 1503, '재구매': 2662, '합니다': 3341, '밥이': 1371, '맛나요': 1005, '맛있어서': 1039, '재구입': 2675, '중간': 2879, '사이즈도': 1667, '있었다면': 2595, '좋았을': 2808, '같아요': 177, '2가지': 53, '뿐이라': 1627, '안성쌀이맛나요ㅡ재구매해요': 2156, '제일맛있는': 2756, '쌀이예요': 2015, '깨진쌀알도없고': 491, '찰지고맛있어요': 2974, '포장': 3193, '배송': 1401, '모두': 1190, '좋네요': 2785, '쌀도': 1944, '믿음이': 1271, '가고': 129, '좋아요': 2796, '앞으로': 2224, '구매하고': 328, '싶습니다': 1909, '빠르게': 1606, '받았어요': 1319, '만세보령쌀': 959, 'gold': 92, '정말': 2743, '맛있습니다': 1036, '먹던쌀이예요': 1108, '아직안뜯어봤는데': 2129, '좋을걸로예상하고있습니다': 2826, '빠른배송': 1609, '감사합니다': 154, '그냥': 401, '좋은데': 2817, '배송잊': 1430, '다른곳으로': 665, '배송문제만': 1412, '빼면': 1620, '괜찮네요': 299, '적극추천': 2710, '배송이': 1429, '빠릅니다': 1611, '빠른': 1608, '보관통이': 1480, '맘에': 1001, '들어요': 861, '등급이': 875, '보통이라': 1523, '신경': 1866, '쓰였지만': 2049, '

In [6]:
# 코사인 유사도 사용
from sklearn.metrics.pairwise import linear_kernel
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

In [7]:
# 리뷰 d와 인덱스를 가진 테이블 생성. 리뷰 아이디를 입력하면 인덱스를 리턴하기 위함
indices = pd.Series(data.index, index=data['id']).drop_duplicates()
print(indices.head())

id
a1    0
a1    1
a1    2
a6    3
a6    4
dtype: int64


In [8]:
idx = indices['a1']
print(idx)

id
a1    0
a1    1
a1    2
dtype: int64


In [17]:
def get_recommendations(title, cosine_sim = cosine_sim):
    idx = indices[id]
    
    # 모든 영화에 대해서 해당 영화의 유사도를 구한다.
    sim_scores = list(enumerate(cosine_sim[idx]))
    
    # 유사도에 따라 영화들을 정렬한다.
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    
    # 가장 유사한 10개의 영화를 받아온다.
    sim_scores = sim_scores[1:11]
    
    # 가장 유사한 10개의 영화의 인덱스를 받아온다.
    kind_indices = [i[0] for i in sim_scores]
    
    # 가장 유사한 10개의 영화의 제목을 리턴한다.
    return data['id'].iloc[kind_indices]

In [18]:
get_recommendations('a1')

IndexError: index 1517305643552 is out of bounds for axis 0 with size 587